### DISCUSSION TITLE
One to One Relationships

### DISCUSSION THREAD

**Discussion by jeremybradbury** [directus/directus#12091] [Feature Requests]
Sometimes we need to pair up 2 tables like user and profile. When we make a 1 to 1, we have the FK in both tables making lookup from either item quick and easy.

In the case of the directus_users table this would be invaluable. Especially if we could pick more than just the id field to join on, why not join on any field with an index?

**Comment by Pachat** [comment]
Hi, did you try?

1. Create a M2O field
2. Click on "Continue in Advanced Field Creation Mode" instead of "Save"
3. Check "Value has to be unique"

**Reply by ched-dev** [reply]
I know this is the suggested solution but I always forget how to do it and think a O2O type could easily be created and supported by default. I'd vote for adding it as a "feature" even if it does something different under the hood.

**Reply by jeremybradbury** [reply]
@Pachat thanks so much for the useful tip, but I'm with @ched-dev here... there is an insane Many to Any feature, but no 1:1. 

M2A?! but no ability to simply add meta fields or several relationships to the M2M tables? That's usually how multiple relationships (aka link tables) are handled... but the strange M2A implementation confuses that pattern out of existence. 

This is literally 1 of 2 key, oft used relationships that are missing. Yet time was spent time making up something new AFA database design instead of providing UX for those.

I am just confused... However, I get that open source gets nutty sometimes. 

But I digress... My point was that this tool can be adapted for a better user experience, as long as maintainer opinions allow.

Thanks again, both of you, for sharing perspectives and knowledge.

**Reply by Pachat** [reply]
I agree: add meta fields to the M2M tables is a feature I miss too.

**Reply by benhaynes** [reply]
I don't see an issue in making a dedicated O2O interface, I presume it would just preconfigure some things like unique in the background. That said, @jeremybradbury, are you saying there are other missing relationships/features? We might want to create other tickets for other requests/issues.

**Reply by benhaynes** [reply]
> I agree: add meta fields to the M2M tables is a feature I miss too.

Unless I'm missing something, this is possible. Just add fields on the junction collection... no?

**Reply by Pachat** [reply]
@benhaynes Interesting. The use case I face is to find the simplest way to sum up multiple evaluations on a form.
Let say we have different subjects: math science geography literature. Each of them has a grade and a comment given by different reviewers (up to 5 reviewers by subject). The fact that they are multiple reviewers for each subject leads to use an M2M.

Ideally, I'd have one collection Grades (student, reviewer, subject)  ( subject is the extra field ) 
Now, how can I have 4 junctions (math / science / geography / literature) to the student form with only one Grades collection?
Ideally, I'd place 4 times the M2M field in the Student form, each time having the subject field hidden with a default value (math / science / geography / literature)

But it looks like it is not possible to create aliases that would point to the same column in the table, each with a different extra value (or I missed something)
and that the "Default value" we see when setting an M2M does not look like to be of some use..

May be there are better designs I don't see?

**Reply by benhaynes** [reply]
I might not have your exact schema in mind, but here's what I'm thinking with a O2M:

**Student**
- id
- grades (o2m)
- etc

**Grades**
- id
- student (fk)
- grade
- subject
- reviewer

**Reply by Pachat** [reply]
@benhaynes  Thank you very much. It helps me to elaborate.

**Reply by jeremybradbury** [reply]
> That said, @jeremybradbury, are you saying there are other missing relationships/features? We might want to create other tickets for other requests/issues.

there are issues using anything but a primary key in many cases, for example i have a unique key in a user table that spans 2 columns that are similar to `external_identifier`  and `provider` and I can use them as foreign keys if i wanted. 

Anything i relate to user is by their primary key only. A lot of disabled fields and default `id`s when you should only need a unique index to make an FK for another table.

i did see that it's possible to add meta fields to M2M tables but they don't show properly or as expected in the admin UI, even after changing permissions etc.

other times when I add a O2M and a matching O2M to the other table, to make a O2O, at least one table shows an FK twice in directus and will have added 2 matching fields to the fields table, and both are required to keep things from breaking.

**Reply by jeremybradbury** [reply]
@Pachat you may actually want that M2A, since you can attach 4 similar tables: math / science / geography / literature. but if you don't have separated tables to start, just use one instead.

so more like like "One Student to Any Grades", just realize it's an extra few queries every time the relationships are traversed, adding some latency with the M2As.

Student

- id
- grades (m2a)
- etc

Math

- id
- student (fk)
- grade (int)
- reviewer (fk)

Sci

- id
- student (fk)
- grade (int)
- reviewer (fk)

Geo

- id
- student (fk)
- grade (int)
- reviewer (fk)

Lit

- id
- student (fk)
- grade (int)
- reviewer (fk)

**Reply by mmmaxxxe** [reply]
This situation would be perfect, but it would be nice to offer the possibility to choose an "insert only" interface (no selection of existing one)
At this point it would be nice to delete orphans in the destination table.

**Comment by jeremybradbury** [comment]
@benhaynes below are some screenshots to describe an example of an issue i am having

Here is the summary

- I have an `external_identifier` in all my users from Twitch (or they create it during the signup process).
- the expectation is this id matches a player, a player is (later) created or field is null for email signup.
  - perhaps eventually we can link up email similarly to other tables with email as their PK or with them as unique values.
- They have a matching `legacy_id` in a profile table (called player here)
- both have unique indexes
- i created a foreign key from player.legacy_id to directus_users.external_id
- when I look in directus or try to make a GraphQL query, it assumes the key is pointed at `id`
- there are no fields in the relations table to specify this field, it only allows id as PKs

![legacy_id to directus_users, via external_id](https://user-images.githubusercontent.com/1664387/160720425-a7a17555-bc73-464e-ba9e-7ae891300cb5.PNG)

![directus forces use of id](https://user-images.githubusercontent.com/1664387/160720542-94d84075-5b8d-406b-a6c9-4ba1c7e31a5b.PNG)

![graphql tries to relate with id instead](https://user-images.githubusercontent.com/1664387/160720581-5158c424-c97a-4205-90a0-be29f5d26386.PNG)

My goal here to allow a user to connect with their Twitch, creating a directus_user with their Twitch ID (not username) in the external_identifier field... then magically have their game data immediately attached to their user via this FK relationship. But RN I'm chasing down triggers or other things to do things that Directus will not.

Later I plan to allow players to also connect their Discord or Google account to the same player / profile record.

But then there is a popular issue here, that I am also working to resolve with my schema: https://github.com/directus/directus/issues/9331


if you still need more things to fix, oauth2 is mostly broken and when I complain they say use OpenID. 

the error reporting for OAuth and OpenID are mostly unusable when errors happen, even with trace enabled.

GraphQL is using 2 endpoints, and that's only useful inside Apollo sever with federation enabled: https://github.com/directus/directus/discussions/12444

there are already tickets/discussions for most of these issues.

like GraphQL over https is now broken https://github.com/directus/directus/issues/12447 I have a PR in to fix it https://github.com/directus/directus/issues/12448 but it's likely just a bandaid backup plan, and another middleware needs to do this much earlier.

but honestly this "PKs are the the only usable unique index" bug for relationships is the biggest deal for me. 

making a clean O2O should be another path to fixing this particular issue in my case, but it then doesn't get us to many accounts per user any time soon.

There's also no way to make FK free relationships (outside of M2A). You use them throughout the app such as fields and relations have no keys or db enforcement. Your users have patterns like these too.

Convention is nice, best practices habits are great... but reality happens too (much of it before adopting Directus) and it's not always pretty. so rather than force the PK for all relationships, you can look at the list of unique indexes and allow the users some choice, with id as the default instead of the assumption.

**Reply by benhaynes** [reply]
I appreciate the comprehensive insight... though I'll have to leave this to the actual engineers. 😉

**Comment by sxsws** [comment]
As this can be hacked together via an M2O, could the O2O field option be added as a readonly textfield with just a create button underneath for the interface and the unique index added automatically? An option to add the reverse O2O could then also be added on the 'child' collection.

It feels odd to not have a O2O field type along with all the other relations.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
